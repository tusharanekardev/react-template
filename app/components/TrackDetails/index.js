/**
 *
 * TrackDetails
 *
 */

import React, { memo, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { connect } from 'react-redux';

// import PropTypes from 'prop-types'
// import styled from 'styled-components'
import PropTypes from 'prop-types';
import { selectItunesData } from '@app/containers/Itunes/selectors';
import { itunesCreators } from '@app/containers/Itunes/reducer';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import CardActionArea from '@mui/material/CardActionArea';
import styled from '@emotion/styled';

const TrackDetailsDiv = styled.div`
  padding: 2rem;
`;

const AudioPlayerDiv = styled.div`
  margin-top: 2rem;
  max-width: 345px;
`;

const CustomAudio = styled.audio`
  width: 100%;
`;

/**
 * TrackDetails component fetches and displays the details of a track using the trackId from the URL parameters.
 *
 * It retrieves data like track name, artist name, collection name, track price, and artwork image from the `tracksData` prop.
 * If no track data is found or available, it triggers a dispatch to fetch the track details by its trackId.
 *
 * @component
 *
 * @param {Object} props - The props passed to the component.
 * @param {Object} props.tracksData - Data containing the list of tracks.
 * @param {Array} props.tracksData.results - The array of track objects containing details.
 * @param {function} props.dispatchItunesTrackById - Function to dispatch an action to fetch track details by ID.
 *
 * @returns {JSX.Element} A card component displaying track details such as name, artist, collection, price, and country.
 *
 * @example
 * // Example usage of TrackDetails component
 * <TrackDetails tracksData={mockTracksData} dispatchItunesTrackById={mockDispatchFunction} />
 */
function TrackDetails({ tracksData, dispatchItunesTrackById }) {
  const { trackId } = useParams();
  const [isAudioPlayer, setIsAudioPlayer] = useState(false);

  const track = tracksData?.results?.find((item) => item?.trackId === parseInt(trackId));

  const {
    artworkUrl100 = '',
    collectionName = 'Unknown Collection',
    artistName = 'Unknown Artist',
    trackName = 'Unknown Track',
    trackPrice = 0,
    country = 'Unknown Country',
    previewUrl = ''
  } = track || {};

  /**
   * Toggles the state of the audio player.
   *
   * This function updates the `isAudioPlayer` state, switching it
   * between `true` and `false`. If `isAudioPlayer` is `true`, the audio
   * player will be visible; if `false`, the audio player will be hidden.
   *
   * @function
   * @returns {void} Does not return a value, but toggles the state internally.
   */
  function handleCardClick() {
    setIsAudioPlayer((prevIsAudioPlayer) => !prevIsAudioPlayer);
  }

  useEffect(() => {
    if (!trackId || !track || !tracksData?.results?.length) {
      dispatchItunesTrackById(trackId);
    }
  }, []);

  return (
    <TrackDetailsDiv data-testid="track-details">
      <Card sx={{ maxWidth: 345 }} data-testid="card-details" onClick={handleCardClick}>
        <CardActionArea>
          <CardMedia component="img" height="140" image={artworkUrl100} alt="green iguana" />
          <CardContent>
            <Typography gutterBottom variant="h5" component="div">
              Track Name: {trackName}
            </Typography>
            <Typography component="p" variant="body1">
              Artist Name: {artistName}
            </Typography>
            <Typography component="p" variant="body1">
              Collection Name: {collectionName}
            </Typography>
            <Typography component="p" variant="body1">
              Track Price: {trackPrice}
            </Typography>
            <Typography component="p" variant="body1">
              Country: {country}
            </Typography>
          </CardContent>
        </CardActionArea>
      </Card>

      {isAudioPlayer && (
        <AudioPlayerDiv data-testid="audio-player">
          <CustomAudio controls>
            <source src={previewUrl} />
          </CustomAudio>
        </AudioPlayerDiv>
      )}
    </TrackDetailsDiv>
  );
}

TrackDetails.propTypes = {
  tracksData: PropTypes.shape({
    resultCount: PropTypes.number,
    results: PropTypes.array
  }),
  dispatchItunesTrackById: PropTypes.func
};

const mapStateToProps = createStructuredSelector({
  tracksData: selectItunesData()
});

/**
 * Maps dispatch actions to component props, enabling the component to dispatch actions to the Redux store.
 *
 * Specifically, this function returns an object containing a function `dispatchItunesTrackById`, which dispatches
 * an action to fetch the iTunes track details by a given trackId.
 *
 * @function
 * @param {function} dispatch - The Redux `dispatch` function used to dispatch actions to the store.
 *
 * @returns {Object} An object containing:
 * - `dispatchItunesTrackById`: A function that takes a trackId and dispatches the `requestGetItunesTrackById` action.
 */
export function mapDispatchToProps(dispatch) {
  const { requestGetItunesTrackById } = itunesCreators;
  return {
    dispatchItunesTrackById: (trackId) => dispatch(requestGetItunesTrackById(trackId))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo)(TrackDetails);
