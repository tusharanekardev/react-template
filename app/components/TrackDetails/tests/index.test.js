/**
 *
 * Tests for TrackDetails
 *
 */

import React from 'react';
import {screen, fireEvent } from '@testing-library/dom';
import { renderProvider} from '@utils/testUtils';
import TrackDetails from '../index';

describe('<TrackDetails />', () => {
  let submitSpy

  beforeEach(() => {
    submitSpy = jest.fn()
  })

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackDetails dispatchItunesTrackById={submitSpy}/>);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackDetails component', () => {
    const { getAllByTestId } = renderProvider(<TrackDetails />);
    expect(getAllByTestId('track-details').length).toBe(1);
  });

  it('should toggle the audio player on button click', () => {
    renderProvider(<TrackDetails />)
    expect(screen.queryByTestId('audio-player')).not.toBeInTheDocument()

    const button = screen.getByTestId('card-details')
    fireEvent.click(button)
    expect(screen.queryByTestId('audio-player')).toBeInTheDocument()

    fireEvent.click(button)
    expect(screen.queryByTestId('audio-player')).not.toBeInTheDocument()
  })
});
