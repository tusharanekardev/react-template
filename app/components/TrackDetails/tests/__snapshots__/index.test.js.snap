// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetails /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-1v03bf2 eh9f6zb2"
      data-testid="track-details"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ri6ub7-MuiPaper-root-MuiCard-root"
        data-testid="card-details"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <img
            alt="green iguana"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            height="140"
          />
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
            >
              Track Name: 
              Unknown Track
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Artist Name: 
              Unknown Artist
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Collection Name: 
              Unknown Collection
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Track Price: 
              0
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Country: 
              Unknown Country
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
